libwacom/libwacom.c

```diff
@@ -1225,7 +1225,7 @@ libwacom_stylus_has_eraser (const WacomStylus *stylus)
 LIBWACOM_EXPORT int
 libwacom_stylus_is_eraser (const WacomStylus *stylus)
 {
-	return stylus->is_eraser;
+	return libwacom_stylus_get_eraser_type(stylus) != WACOM_ERASER_NONE;
```

Comment:
I think this should only return true where the tool itself is only ever an eraser, i.e. for `WACOM_ERASER_INVERT`. That seems to be the most reliable backwards-compatibility map.

Reply:
That sounds fine to me. How do we want `libwacom_stylus_has_eraser` to behave? Right now it only returns true if a stylus is paired with an eraser, but is not itself one. Do we want it to also return true if the stylus may _become_ an eraser (i.e., is an AES pen)? Alternatively, do we want to introduce a new function (e.g. `libwacom_stylus_can_erase`) that would be true for an AES pen?

Whatever we do, those functions could use some documentation that clarifies the difference between "is" and "has" (and possibly "can") :wink:

Reply:
I'd say exact same behaviour as now (i.e. return `false`) and deprecate it. A quick check at the gnome-control-center shows that the `has-eraser` flag is used in combination with the button count to determine the "layout" of the panel. from libwacom's POV that is outside our control so the safest thing is to not break anything.

gnome is fast-moving enough that this won't be an issue for long anyway.


> do we want to introduce a new function (e.g. libwacom_stylus_can_erase)

yes, but why not go for the same approach with `get_eraser_type()` and return an enum value? More future-proof IMO.

Reply:
The more I look at this, the more I think compatibility really isn't an issue and the more I prefer the implementation under review. As far as I can tell, the behavior _is_ backwards compatible:

* Callers expecting the current behavior will continue to function correctly with the various invert-style erasers that we may introduce in the future.
* Callers expecting future behavior will work correctly with current and past versions of libwacom for the styli that are in their databases (which are all invert-style).

The only compatibility argument that I can see is that callers which expect current behavior may mistreat the new button-style erasers as invert-style. I'm not concerned about this, however, since those callers would need other updates (even _if_ we changed the API) to properly support button-style erasers.

Reply:
unless I'm misunderstanding something, my worry is about *current callers* with *updated libwacom*. Right now, we have pens that **have** an eraser and pens that **are** an eraser (but those have a different ID). In the case of AES pens, I don't think they change the ID for the eraser, so with the check for `!= NONE` we now basically say that every AES pen is an eraser pen and there are no stylus pens anymore.

I still think it'd be better to fail the other way around, so older applications will think every AES pen is a stylus without an eraser.

Reply:
It seems just as wrong for an application to treat an AES pen as a having no eraser as it is to treat the pen as *being* the eraser. Your AES pen goes from "doesn't work at all" with the current libwacom release to "works incorrectly" either way. The only way for an application to get to "works correctly" is to start using `libwacom_stylus_get_eraser_type` and write new code to handle button-style erasers.

The reason I prefer `!= NONE` comes down to trying to find an eraser given an arbitrary tool ID. With this definition I can simply call `is_eraser` on each tool and know when I've found an ID that can erase. Using `== WACOM_ERASER_INVERT` instead makes `is_eraser` useless for me, and so I end up write my own `!= NONE` check anyway.

Reply:
both are wrong, but I'd say one is less so than the other. not having a working eraser button should be fine for a large fraction of use-cases [1] esp given that there are also calls for mapping the eraser button to just be a stylus button. So right now, having the eraser button not work doesn't bother me as much as having the stylus potentially not working.

> Using `== WACOM_ERASER_INVERT` instead makes is_eraser useless for me, and so Iend up write my own != NONE check anyway.

that's true, but only for the simplest client. If we assume that a client is interested in whether an eraser is a button vs pen, then that doesn't apply anymore because you need a custom check anyway. And you'd be using  `libwacom_stylus_get_eraser_type` instead.
Note that libwacom isn't a framework, we have what - 2 or 3 actual users with little to no churn in the code that calls our API. So even if it is a few more lines for those they get written once and then never touched again.


[1] and we're just talking about backwards-compat anyway.

Reply:
Sorry for the slow rate of replies. Been bouncing between a bunch of different tasks for the past several weeks :(

I'm still not really buying this argument. You might not be able to configure your AES pen because the tool thinks its an eraser, but its still going to work. The worst case scenario I can think of is a caller monitoring the event stream and using the result of `is_eraser` (rather than the event's own stylus/eraser flag) to tell applications "you should erase with this tool". None of our "2 or 3 actual users" have ever done done this, and I doubt there are any others out there that do. Libwacom is primarily used for configuration after all. In reality you just won't be able to properly configure your AES pen, which isn't any different from today.

> that's true, but only for the simplest client. If we assume that a client is interested in whether an eraser is a button vs pen, then that doesn't apply anymore because you need a custom check anyway. And you'd be using libwacom_stylus_get_eraser_type instead.

I for one found it handy to have them around as utility functions within libwacom itself. I imagine you wouldn't *always* care about the specific eraser type (especially in the `has_eraser` case), and these functions cater exactly to those situations. I don't see much sense in requiring callers to re-implement their functionality in order to avoid a compatibility issue which probably doesn't exist.

Reply:
Any feedback, @whot?